storage/kafka/datasource.go

```diff
@@ -0,0 +1,349 @@
+			}
+			continue
+		} else if err != nil {
+			tx.Abort()
+			log.Printf("kafka worker: error running kafka read messages worker: %s, reinitializing from storage", err)
+			tx := rs.stateStorage.BeginTransaction().WithPrefix(rs.streamID.AsPrefix())
+			if err := rs.loadOffset(tx); err != nil {
+				log.Fatalf("kafka worker: couldn't reinitialize offset for kafka read messages worker: %s", err)
+				return
+			}
+			tx.Abort() // We only read data above, no need to risk failing now.
+			continue
+		}
+
+		err = tx.Commit()
+		if err != nil {
+			log.Println("kafka worker: couldn't commit transaction: ", err)
+			continue
+		}
+	}
+}
+
+func (rs *RecordStream) RunWorkerInternal(ctx context.Context, tx storage.StateTransaction) error {
+	outputQueue := execution.NewOutputQueue(tx.WithPrefix(outputQueuePrefix))
+	tokenQueue := execution.NewOutputQueue(tx.WithPrefix(tokenQueuePrefix))
+
+	// The worker needs a token to read a batch.
+	var token octosql.Value
+	err := tokenQueue.Pop(ctx, &token)
+	if err != nil {
+		return errors.Wrap(err, "couldn't get batch token from token queue")
+	}
+
+	batch := make([]*execution.Record, rs.batchSize)
+	for i := 0; i < rs.batchSize; i++ {
+		msg, err := rs.kafkaReader.ReadMessage(ctx)
+		if err != nil {
+			return errors.Wrap(err, "couldn't read message from kafka")
+		}
+
+		fields := []octosql.VariableName{octosql.NewVariableName(fmt.Sprintf("%s.key", rs.alias))}
+		values := []octosql.Value{octosql.MakeString(string(msg.Key))}
+		if !rs.decodeAsJSON {
+			fields = append(fields, octosql.NewVariableName(fmt.Sprintf("%s.value", rs.alias)))
+			values = append(values, octosql.MakeString(string(msg.Value)))
+		} else {
+			object := make(map[string]interface{})
+			err := json.Unmarshal(msg.Value, &object)
+			if err != nil {
+				log.Printf("couldn't decode value as json: %s, passing along as raw bytes", err)
+				fields = append(fields, octosql.NewVariableName("value"))
+				values = append(values, octosql.MakeString(string(msg.Value)))
+			} else {
+				for k, v := range object {
+					fields = append(fields, octosql.NewVariableName(fmt.Sprintf("%s.%s", rs.alias, k)))
+					values = append(values, octosql.NormalizeType(v))
+				}
+			}
+		}
+		batch[i] = execution.NewRecordFromSlice(fields, values)
+	}
+
+	for i := range batch {
+		err := outputQueue.Push(ctx, &QueueElement{
+			Type: &QueueElement_Record{
+				Record: batch[i],
+			},
+		})
+		if err != nil {
+			return errors.Wrapf(err, "couldn't push kafka message with index %d in batch to output record queue", i)
+		}
+	}
+
+	if err := rs.saveOffset(tx); err != nil {
+		return errors.Wrap(err, "couldn't save kafka offset")
+	}
+
+	return nil
+}
+
+func (rs *RecordStream) loadOffset(tx storage.StateTransaction) error {
+	offsetState := storage.NewValueState(tx.WithPrefix(offsetPrefix))
+
+	var offset octosql.Value
+	err := offsetState.Get(&offset)
+	if err == storage.ErrNotFound {
```

Comment:
I don't know how offset works in kafka reader but shouldn't we save some 0 offset?

Reply:
the kafka reader will default to something, but I decided to additionally add a startOffset config variable.